File: bench-core/node/provenance.mjs
// Provenance attached to every result so a number is traceable to a run: timestamp,
// git commit, machine/OS/node, and the harness version. node-only.
import { execSync } from 'node:child_process';
import os from 'node:os';

export const HARNESS_VERSION = '@viz-bench/bench-core/1';

export function provenance({ harness = HARNESS_VERSION } = {}) {
  let gitCommit = null;
  try { gitCommit = execSync('git rev-parse --short HEAD', { encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] }).trim(); } catch {}
  let machine = process.arch;
  try { machine = os.cpus()?.[0]?.model || process.arch; } catch {}
  return {
    timestamp: new Date().toISOString(),
    gitCommit,
    machine,
    os: `${process.platform} ${os.release()}`,
    node: process.version,
    harness,
  };
}

File: bench-core/py/provenance.py
"""Provenance for native (Python) bench records — mirrors bench-core/node/provenance.mjs."""
import subprocess, platform, sys, datetime

HARNESS_VERSION = "@viz-bench/bench-core/1-py"


def _machine():
    try:
        s = subprocess.run(["sysctl", "-n", "machdep.cpu.brand_string"], capture_output=True, text=True)
        return s.stdout.strip() or platform.machine()
    except Exception:
        return platform.machine()


def provenance(harness=HARNESS_VERSION):
    try:
        git = subprocess.run(["git", "rev-parse", "--short", "HEAD"], capture_output=True, text=True).stdout.strip() or None
    except Exception:
        git = None
    return {
        "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "gitCommit": git,
        "machine": _machine(),
        "os": f"{sys.platform} {platform.release()}",
        "python": sys.version.split()[0],
        "harness": harness,
    }
